src/main/java/com/pingcap/tikv/types/DataType.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2017 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pingcap.tikv.types;
+
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.pingcap.tikv.codec.CodecDataInput;
+import com.pingcap.tikv.codec.CodecDataOutput;
+import com.pingcap.tikv.meta.Collation;
+import com.pingcap.tikv.row.Row;
+import com.pingcap.tikv.meta.TiColumnInfo;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.pingcap.tikv.types.Flags.*;
+
+/**
+ * Base Type for encoding and decoding TiDB row information.
+ */
+public class DataType {
+     public enum EncodeType {
+        KEY,
+        VALUE
+    }
+    public static final int UNSPECIFIED_LEN = -1;
+
+    // MySQL type
+    protected int           tp;
+    // Not Encode/Decode flag, this is used to strict mysql type
+    // such as not null, timestamp
+    protected int           flag;
+    protected int           collation;
+    protected int           length;
+    private   List<String> elems;
+
+
+   static final Map<Integer, String> codecMap = ImmutableMap.<Integer, String>builder()
```

Comment:
either make back to subclass interface or add comment in type codec flag declare.
For current design when you add codec will make name translate throw exception.